### PR TITLE
WIP: Update tagger model (char2tag)

### LINKED ISSUE #13: Compatibility with Python 3
Finally got round to experimenting with Mimick only to discover that it targets Python 2 only. (*Insert rant that Python 3 is already a decade old.*) Do you by any chance plan to add support for Python 3?

Thanks!

### LINKED ISSUE #15: Char2Tag takes wrong representations from backward LSTM
[This line](https://github.com/yuvalpinter/Mimick/blob/f781e15aaec5a9e6b77704adcf548d4f38628c04/model.py#L110) should not be concatenating `char_embs[-1]`, but rather `dy.concatenate([char_embs[-1][:h],char_embs[0][h:]])` for the appropriate `h`.

The in-model Mimick code is fine, since it uses separate fwd and bwd char-level models rather than dynet's built-in `BiRNNBuilder`. The word-level BiLSTM is fine because it's sequence prediction.

### LINKED ISSUE #16: Add early stopping

### LINKED ISSUE #17: Remove `START_TAG` and `STOP_TAG` from `model.py` and `make_dataset.py`
They're dead references, back from the ancient version of the code that had a CRF w/ Viterbi decoding on top of the word-leve LSTM.

### PR DESCRIPTION
**Pull request by ruyimarone**
- resolves #17
- resolves #15
- resolves #16 (early stopping) for the char2tag model (keeps the model with the highest dev POS tagging accuracy) 
- resolves #13 (python3 compatibility) for char2tag related files (`model.py`, `utils.py`, `make_dataset.py`, `evaluate_morphotags.py`)

Still WIP pending tests of new models, and updating the rest of Mimick to python3 so the full repo is compatible.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -177,41 +180,17 @@ def set_dropout(self, p):
...
-        members_to_save.extend(utils.sortvals(self.lstm_to_tags_params))
-        members_to_save.extend(utils.sortvals(self.lstm_to_tags_bias))
-        members_to_save.extend(utils.sortvals(self.mlp_out))
-        members_to_save.extend(utils.sortvals(self.mlp_out_bias))
-        self.model.save(file_name, members_to_save)
 
         with open(file_name + "-atts", 'w') as attdict:
             attdict.write("\t".join(sorted(self.attributes)))
 
-    @property
```
Comment (yuvalpinter) [inline model.py:211]: is this function's deletion due to dynet 2.0 or something else?

File: model.py [hunk 2]
```diff
@@ -356,7 +336,10 @@ def get_word_chars(sentence, i2w, c2i):
     logging.info("Number training instances: {}".format(len(training_instances)))
     logging.info("Number dev instances: {}".format(len(dev_instances)))
 
-    for epoch in xrange(int(options.num_epochs)):
+    best_dev_pos = 0.0
+    old_best_name = None
+
+    for epoch in range(int(options.num_epochs)):
```
Comment (yuvalpinter) [inline model.py]: is this int casting still required? maybe make the `num_epochs` param into an `int` value?

File: model.py [hunk 3]
```diff
@@ -431,14 +414,14 @@ def get_word_chars(sentence, i2w, c2i):
...
-                total_loss = sum([l.scalar_value() for l in losses.values()])
+                total_loss = sum([l.scalar_value() for l in list(losses.values())])
                 out_tags_set = model.tag_sentence(instance.sentence, word_chars)
 
                 gold_strings = utils.morphotag_strings(i2ts, gold_tags)
                 obs_strings = utils.morphotag_strings(i2ts, out_tags_set)
                 for g, o in zip(gold_strings, obs_strings):
                     f1_eval.add_instance(utils.split_tagstring(g, has_pos=True), utils.split_tagstring(o, has_pos=True))
-                for att, tags in gold_tags.items():
+                for att, tags in list(gold_tags.items()):
```
Comment (yuvalpinter) [inline model.py:424]: pretty sure this change isn't necessary (order doesn't matter), same for l. 537

File: model.py [hunk 4]
```diff
@@ -491,6 +475,22 @@ def get_word_chars(sentence, i2w, c2i):
             old_devout_file_name = "{}/devout_epoch-{:02d}.txt".format(options.log_dir, epoch)
             os.remove(old_devout_file_name)
 
+        # write best model by dev pos accuracy in addition to periodic writeouts
+        if dev_pos_accuracy > best_dev_pos:
+            logging.info("{:.4f} > {:.4f}, writing new best dev model".format(dev_pos_accuracy * 100,
```
Comment (yuvalpinter) [inline model.py:480]: now that it's python 3, this can be f-string fun instead (l. 490, 496 as well)

File: model.py [hunk 5]
```diff
@@ -505,6 +505,7 @@ def get_word_chars(sentence, i2w, c2i):
         # epoch loop ends
 
     # evaluate test data (once)
+    #TODO this should be done using the best dev model
```
Comment (yuvalpinter) [inline model.py:508]: open issue if not resolved by final PR version

**Review by yuvalpinter — COMMENTED** [review]
looks great overall, see comments, and let me know when it's ready to be merged.
